internal/controller/bmc/controller.go

```diff
@@ -0,0 +1,333 @@
+// +kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=kubevirt.io,resources=virtualmachines,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=secrets,verbs=get;list;watch
+
+const (
+	ConditionReady    = "Ready"
+	ConditionNotReady = "NotReady"
+
+	ReasonReady  = "AllComponentsReady"
+	MessageReady = "VirtualMachineBMC is ready"
+
+	VirtualMachineNotFound        = "VirtualMachineNotFound"
+	MessageVirtualMachineNotFound = "Referenced VirtualMachine does not exist"
+
+	AuthSecretNotFound        = "AuthSecretNotFound"
+	MessageAuthSecretNotFound = "Referenced Secret does not exist"
+)
+
+var (
+	ErrVirtualMachineBMCNil   = errors.New("virtualMachineBMC is nil")
+	ErrVirtualMachineRefNil   = errors.New("virtualMachineRef is nil or empty")
+	ErrAuthSecretRefNil       = errors.New("authSecretRef is nil or empty")
+	ErrVirtualMachineNotFound = errors.New("referenced virtualMachine not found")
+	ErrAuthSecretNotFound     = errors.New("referenced secret not found")
+)
+
+func (r *VirtualMachineBMCReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.Log.WithValues("VirtualMachineBMC", req.NamespacedName).Info("Reconciling VirtualMachineBMC")
+
+	var virtBMC bmcv1beta1.VirtualMachineBMC
+	if err := r.Get(ctx, req.NamespacedName, &virtBMC); err != nil {
+		r.Log.Error(err, "Failed to get VirtualMachineBMC")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	res, err := r.ValidateRefs(ctx, &virtBMC)
+	if err != nil || res.Requeue {
+		if errors.Is(err, ErrVirtualMachineNotFound) || errors.Is(err, ErrAuthSecretNotFound) {
+			var (
+				errorReason  = AuthSecretNotFound
+				errorMessage = MessageAuthSecretNotFound
+			)
+
+			if errors.Is(err, ErrVirtualMachineNotFound) {
+				errorReason = VirtualMachineNotFound
+				errorMessage = MessageVirtualMachineNotFound
+			}
+			r.Log.Error(err, "Referenced resource not found, setting NotReady condition", "Reason", errorReason, "Message", errorMessage)
+			cond := metav1.Condition{
+				Type:               ConditionNotReady,
+				Status:             metav1.ConditionFalse,
+				Reason:             errorReason,
+				Message:            errorMessage,
+				ObservedGeneration: virtBMC.Generation,
+			}
+			setCondition(&virtBMC, cond)
+			if updateErr := r.Status().Update(ctx, &virtBMC); updateErr != nil {
+				r.Log.Error(updateErr, "Failed to update VirtualMachineBMC status")
+				return ctrl.Result{}, updateErr
+			}
+			r.Log.Info("Cleaning up resources due to missing references")
+			if err := r.DeleteResources(ctx, &virtBMC); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			r.Log.Info("Resources cleaned up due to missing references")
+			return ctrl.Result{Requeue: true}, nil
```

Comment:
Why requeue it now?

Reply:
I want retry reconciliation in case the missing references (e.g., VM or Secret) come back as soon as possible.
Do you think it is redundant ?